### DISCUSSION TITLE
hydra-zen: Building a bridge between Hydra and pydantic

### DISCUSSION THREAD

**Discussion by rsokl** [pydantic/pydantic#3751] [General]
Hi all!

This isn't a request or a question; it is really a show-and-tell 😄.

I am the lead dev on [hydra-zen](https://github.com/mit-ll-responsible-ai/hydra-zen), which aims to make [Hydra](https://hydra.cc/) much easier to use and to enable novel capabilities in the framework.  I believe that some of hydra-zen's function make it so that pydantic and Hydra can be used together! I am hoping to get some feedback from folks on the pydantic team.

## Some Context About hydra-zen

 The "crown jewel" of `hydra-zen` is the function [`hydra_zen.builds`](https://mit-ll-responsible-ai.github.io/hydra-zen/generated/hydra_zen.builds.html), which automatically generates structured configs (i.e. dataclasses that Hydra understands):

<table>

<tr>
<td> Using Plain Hydra </td> <td> Using Hydra-Zen </td>
</tr>

<tr>
<td>

```python
from dataclasses import dataclass, field

def foo(bar: int, baz: list[str], qux: float = 1.23):
    ...

@dataclass
class FooConf:
    _target_: str = "__main__.foo"
    bar: int = 2
    baz: list[str] = field(default_factory=lambda: ["abc"])
    qux: float = 1.23
```
<td/>

```python
from hydra_zen import builds

def foo(bar: int, baz: list[str], qux: float = 1.23):
    ...

FooConf = builds(foo, bar=2, baz=["abc"], populate_full_signature=True)



```
</tr>

</table>

Thus `builds` has the ability to inspect the signature of the target-object that we want to configure, and to create a Hydra-compatible dataclass that configures it. It also has many bells-and-whistles that keep users from snagging on any of Hydra's sharp edges. Most relevant to this discussion: `builds` will [automatically widen type-annotations](https://mit-ll-responsible-ai.github.io/hydra-zen/explanation/type_refinement.html) that are not supported by Hydra's type-validation system.

## Using pydantic models in Hydra

So, how does `builds` provide a bridge between pydantic and Hydra? Let's check out an example:

```python
from pydantic import BaseModel, PositiveFloat

from hydra_zen import builds, instantiate

import hydra
from hydra.core.config_store import ConfigStore


# The interface to our Hydra-app will be defined by a pydantic model!
class PydConf(BaseModel):
    val: PositiveFloat 


# Create Hydra-compatible dataclass that describes `PydConf`
HydraConf = builds(PydConf, populate_full_signature=True)

cs = ConfigStore.instance()
cs.store(name="pydantic_conf", node=HydraConf)


# Our CLI is created to reflect the signature of `PydConf`
@hydra.main(config_path=None, config_name="pydantic_conf")
def task_function(cfg: HydraConf):
    obj: PydConf = instantiate(cfg)  # `calls PydConf(val=cfg.val)``
    
    assert isinstance(obj, PydConf)
    print(repr(obj))


if __name__ == "__main__":
    task_function()
```

We can then call our Hydra-based application from the command line, and get the benefit of our pydantic model's validation:

```shell
$ python pyd.py val=1.0
PydConf(val=1.0)

$ python pyd.py val=-2.0
<pydantic ValidationError raised>
```

Thus we've gained all the benefits of using the Hydra framework while also leveraging pydantic's type coercion and validation abilities! 

### How does this work?

Let's understand what is going on here; `builds(PydConf, populate_full_signature=True)` creates a dataclass whose signature matches that of `PydConf`, but whose type-annotations have been widened, where necessary, to appease Hydra's limited type-validation system. Thus `val: PositiveFloat` -> `val: Any`. This is fine though, because pydantic will ultimately be performing type-validation via `PydConf`; `HydraConf` is just a temporary "vessel" that defines the outer-most interface to our Hydra application/CLI.

Within our task-function, we use Hydra's `instantiate` API to instantiate all of the objects that we have described with our configs. E.g. `instantiate(HydraConf, val=1.0)` ends up calling `PydConf(val=1.0)` and returns the result. Thus we ultimately instantiate the actual pydantic model before proceeding with the rest of our application.

## There's a (temporary) catch...

The Hydra-framework attempts to catch and re-raise errors with modified error messages, however `pydantic.ValidationError` has an atypical signature, and so Hydra's attempt to call `pydantic.pydantic.ValidationError(new_err_msg)` produces a `TypeError`. Thus the validation error raised in the above example gets obscured.

Fortunately [Hydra has merged a fix for this](https://github.com/facebookresearch/hydra/pull/1916) for their next minor release. (I also wonder if it might be worth refactoring `ValidationError` to accommodate this common usage.)

## One other way to use pydantic with Hydra

`hydra_zen.builds` provides a generic ability to auto-apply (one or more) wrappers to any target object being configured/instantiated. Thus `pydantic.validate_arguments` can be auto-applied to any/all interfaces being configured in your application:

```python
from hydra_zen.third_party.pydantic import validates_with_pydantic
from hydra_zen import make_custom_builds_fn, instantiate

validated_builds = make_custom_builds_fn(
    zen_wrappers=validates_with_pydantic, populate_full_signature=True
)

# Any config created using `validated_builds(<target>)` will instantiate as `validate_arguments(<target>)(*args, **kwargs)`
```

```python
from typing_extensions import TypedDict

class UserIdentity(TypedDict):
    name: str

def some_func(x: UserIdentity):
    return x

Config = validated_builds(some_func)
```

```python
instantiate(Config, x={"not_a_name": "hello"})  # raises pydantic.ValidationError

instantiate(Config, x={"name": "Ryan"})  # OK
```

## Feedback

I'd love to get your thoughts on the above.

- Does this seem interesting/useful to you?
- Are there any pitfalls that you can anticipate?
- Would you be interested in eventually including a section in your docs that demonstrate this usage of pydantic within a Hydra-based application?

Thanks for taking the time to read through this!

**Comment by esafak** [comment]
Is it not possible to use Hydra with pydantic without hydra-zen? When I use a pydantic type in a pydantic dataclass with Hydra I get a `ValidationError` like `Unexpected object type: ConstrainedIntValue`. I also get warnings about the `_target_` field:
> fields may not start with an underscore, ignoring `"_target_"`

 If it is possible I would appreciate an example.

**Reply by rsokl** [reply]
No, there is no straightforward way to use Hydra without hydra-zen. This is due to [Hydra's limited support for type annotations](https://mit-ll-responsible-ai.github.io/hydra-zen/explanation/type_refinement.html).

Fortunately, it is pretty trivial to use hydra-zen, it basically adds two lines of code in order to "bridge" your pydantic config with your Hydra app. I gave an example above, but here is the core of it:

```python
from pydantic import BaseModel, PositiveFloat

from hydra_zen import builds


class PydConf(BaseModel):
    val: PositiveFloat 


# Create Hydra-compatible dataclass that describes `PydConf`
HydraConf = builds(PydConf, populate_full_signature=True)

# pass `HydraConf` to your Hydra app, rather than `PydConf`
```

I am not sure what your use case is; let me know if this doesn't clarify things for you.

**Reply by esafak** [reply]
Thanks for the explanation. Unfortunately, this does not seem to support `_target_` in my tests and loses [code completion in PyCharm](https://www.jetbrains.com/help/pycharm/auto-completing-code.html). When I run `print(hydra_zen.to_yaml(cfg))` after instantiation I find `_target_` points to the dataclass rather than the method.

**Reply by rsokl** [reply]
> this does not seem to support `_target_`

I am not sure what you mean by this. You can access the `_target_` attribute on the dataclass that is produced by `builds`:

```python
>>> from hydra_zen import builds
>>> builds(int)._target_
'builtins.int'
```

> loses [code completion in PyCharm](https://www.jetbrains.com/help/pycharm/auto-completing-code.html)

Once you instantiate the config within the your app's task function your IDE will "see" that it is `PydConf` instance, and will autocomplete accordingly:

![image](https://user-images.githubusercontent.com/29104956/153611781-8edc1d18-8e54-4300-a761-99a2abb375c7.png)


In the case where you want to use a pydantic model to define an interface in a Hydra app, `builds` is only meant to provide an intermediate Hydra-compatible dataclass.

> When I run `print(hydra_zen.to_yaml(cfg))` after instantiation I find `_target_` points to the dataclass rather than the method.

Yes, this is as-expected (although, I am not sure what method you are referring to). `builds(PydConf, populate_full_signature=True)` creates a structured config (i.e. a dataclass with a `_target_` field) that describes how to "build" `PydConf`. Thus it is effectively equivalent to:

```python
@dataclass
class HydraConf:
    _target_: str = 'main.PydConf'
    val: Any
```

The idea here is that this dataclass only uses types that are compatible with Hydra, however when you actually instantiate the config, the original pydantic model will be instantiated, and thus will trigger runtime type-checking.

```python 
instantiate(HydraConf, val=2.0)  # equivalent to calling `PydConf(val=2.0)`
```


It would be useful for me to see an example of the code you are writing.

**Reply by rsokl** [reply]
@esafak I realized that I can improve the code-completion for `builds(<target>, populate_full_signature=True)`. I have a WIP PR open: https://github.com/mit-ll-responsible-ai/hydra-zen/pull/224

**Reply by esafak** [reply]
Thanks for the prompt and detailed response! Here is a minimal example, which does not call `foo` as expected:

```python
import hydra
import hydra_zen
import pydantic
from hydra.core.config_store import ConfigStore
from hydra_zen import make_custom_builds_fn, hydrated_dataclass
from hydra_zen.third_party.pydantic import validates_with_pydantic

validated_builds = make_custom_builds_fn(zen_wrappers=validates_with_pydantic, populate_full_signature=True)


def foo(value):
    print('We got', value)


class _Task(pydantic.BaseModel):
    value: pydantic.PositiveInt
    _target_: str = '__main__.foo'


Task = hydra_zen.builds(_Task, populate_full_signature=True)


@hydra.main(config_path='', config_name="task")
def init(cfg: Task):
    hydra_zen.instantiate(cfg)


if __name__ == "__main__":
    cs = ConfigStore.instance()
    cs.store(name='task', node=Task)
    init()
```
Also, I find having to essentially define Task twice inelegant. Could we not use a decorator? And since I'm asking, is it possible to define the target function in terms of its associated dataclass; i.e., having one dataclass argument, rather than as many arguments as the dataclass has fields? That would be cool!

# Code completion
The dataclass field `value` is missing. This is important because the Task is supposed to transparently pass for a dataclass, and expose its fields. I will not be building or instantiating it elsewhere in the code base where I will want to re-use this dataclass; only here.

<img width="585" alt="Look ma, no fields!" src="https://user-images.githubusercontent.com/3928300/153651779-7b37f1cb-9190-4b38-85bf-ba891c786e12.png">

**Reply by rsokl** [reply]
> Also, I find having to essentially define Task twice inelegant.

You don't need your `_Task` class at all! Instead, just annotate `foo` with the types you desire and pass it to `builds` directly. This is the point of `builds`: it dynamically generates the structured config for you, so that you don't have to replicate all of your function interfaces by-hand in your configs.

```python
import hydra
import hydra_zen
import pydantic
from hydra.core.config_store import ConfigStore
from hydra_zen import make_custom_builds_fn
from hydra_zen.third_party.pydantic import validates_with_pydantic

validated_builds = make_custom_builds_fn(zen_wrappers=validates_with_pydantic, populate_full_signature=True)

def foo(value: pydantic.PositiveInt):
    print('We got', value)

Task = validated_builds(foo)

@hydra.main(config_path='', config_name="task")
def init(cfg: Task):
    hydra_zen.instantiate(cfg)


if __name__ == "__main__":
    cs = ConfigStore.instance()
    cs.store(name='task', node=Task)
    init()
```

```shell
$ python my_app.py value=2
We got 2

$ python my_app.py value=-2
<Error>
```

>  is it possible to define the target function in terms of its associated dataclass; i.e., having one dataclass argument, rather than as many arguments as the dataclass has fields? That would be cool!

Could you clarify this? I am not sure I follow.

### Code Completion

Barring writing IDE-specific plugins, I do not think it is possible to write a function that returns a dynamically-generated dataclass whose attributes are "discoverable" by the IDE.


```python
Task = validated_builds(foo)

Task. # <- there is no way for the IDE to "know" that .value is available as an attribute
```

That being said, hydra-zen does strive to provide as much type info as possible:

![image](https://user-images.githubusercontent.com/29104956/153658052-ee35b448-638f-4f95-906c-76fd257edf48.png)

We do provide a [special dataclass-like decorator](https://mit-ll-responsible-ai.github.io/hydra-zen/explanation/hydrated_dataclass.html#combining-statically-defined-and-dynamically-generated-configurations) for combining static and runtime checks, but this cannot be used with pydantic's types.

**Reply by esafak** [reply]
I dislike generating the dataclass from the function for several reasons:

1. The data is primal; functions merely act on it. This is reflected in the fact that my real dataclasses are composed; I can share  common parameters between dataclasses using another dataclass.
2. It is not obvious to readers that `validated_builds` yields a dataclass.
3. Code completion for the generated dataclass does not work, as discussed.

I would rather pass the dataclass to the function:
```python
def foo(data: MyDataclass):
 pass
```

**Reply by rsokl** [reply]
> The data is primal; functions merely acts on it. This is reflected in the fact that my real dataclasses are composed; I can share common parameters between dataclasses using another dataclass.

hydra-zen's perspective on this is that your function's signature should be "primal". If you consider configuring functions from a third-party library, you wouldn't want to replicate all of their interfaces with hand-written dataclasses. This violates [DRY](https://mit-ll-responsible-ai.github.io/hydra-zen/explanation/dont_repeat_yourself.html).

> I can share common parameters between dataclasses using another dataclass.

`builds` supports inheritance, so you can share common parameters through it as well.

I suspect that there is probably a way you can use hydra-zen to make Hydra work with pydantic for your use case -- using some utility function you could write to facilitate it. Let me know if you figure anything out to satisfy your use case. Best of luck! 😄

**Comment by rsokl** [comment]
Update: https://mit-ll-responsible-ai.github.io/hydra-zen/how_to/pydantic_guide.html
